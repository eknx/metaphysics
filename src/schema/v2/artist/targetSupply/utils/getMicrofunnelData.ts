import { flatten, groupBy } from "lodash"
import { staticCSVToJSONData } from "./staticCSVToJSONData"

interface ArtistConsignment {
  artworks: Array<{
    internalID: string
    realizedPrice: string
    [key: string]: string
  }>
  metadata: {
    highestRealized: string
    realized: string
    recentlySoldArtworkIDs: string[]
    roundedUniqueVisitors: string
    roundedViews: string
    str: string
    uniqueVisitors: string
    views: string
  }
}

/**
 * Massage CSVtoJSON data to be more usable
 */
export function getMicrofunnelData(pathname: string): ArtistConsignment {
  const dataGroupedByURL = groupBy(staticCSVToJSONData, "url")

  const mappedData = Object.entries(dataGroupedByURL).reduce(
    (acc, [key, value]) => {
      /**
       * The first item in the array contains metrics related to artist; the rest
       * are artworks to display
       */
      const [meta, ...artworks] = value

      const recentlySoldArtworkIDs = flatten(
        artworks.map((artwork: any) => {
          const id = artwork["Artwork ids (recently sold) (comma separated)"]
          const realizedPrice = artwork["Realized Price (in dollars)"]

          // Create aliases that map to GraphQL types
          artwork.internalID = id
          artwork.realizedPrice = realizedPrice
          return id
        })
      )

      // Make CSV column metadata easier to reference
      const {
        "views to content from that artist (last month)": views,
        "ROUNDED views to content from that artist (last month)": roundedViews,
        "unique visitors to content from artist (last month)": uniqueVisitors,
        "ROUNDED unique visitors to content from artist (last month)": roundedUniqueVisitors,
        "STR (last 12 months)": str,
        "Realized / estimate": realized,
        "Highest realized price": highestRealized,
      } = meta

      return {
        ...acc,
        [key]: {
          artworks,
          metadata: {
            views,
            roundedViews,
            uniqueVisitors,
            roundedUniqueVisitors,
            str,
            realized,
            highestRealized,
            recentlySoldArtworkIDs,
          },
        },
      }
    },
    {}
  )

  const microfunnelData = mappedData[pathname]
  return microfunnelData
}

export function getMicrofunnelDataByArtworkInternalID(internalID: string) {
  const dataGroupedByInternalID = groupBy(
    staticCSVToJSONData,
    "Artwork ids (recently sold) (comma separated)"
  )
  const microfunnelData = dataGroupedByInternalID[internalID]
  return microfunnelData?.[0]
}
