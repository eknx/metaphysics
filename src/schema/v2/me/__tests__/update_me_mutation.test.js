/* eslint-disable promise/always-return */
import { runAuthenticatedQuery } from "schema/v2/test/utils"
import gql from "lib/gql"

describe("UpdateMeMutation", () => {
  it("updates the user profile and returns its new data payload", async () => {
    const mutation = gql`
      mutation {
        updateMyUserProfile(
          input: {
            collectorLevel: 1
            clientMutationId: "1232"
            phone: "1234890"
            location: { address: "123 my street" }
            priceRangeMin: -1
            priceRangeMax: 1000000000000
            receivePurchaseNotification: false
            receiveOutbidNotification: false
            receiveLotOpeningSoonNotification: false
            receiveSaleOpeningClosingNotification: false
            receiveNewWorksNotification: true
            receiveNewSalesNotification: false
            receivePromotionNotification: false
          }
        ) {
          user {
            name
            phone
            location {
              city
              address
            }
            priceRange
            receivePurchaseNotification
            receiveOutbidNotification
            receiveLotOpeningSoonNotification
            receiveSaleOpeningClosingNotification
            receiveNewWorksNotification
            receiveNewSalesNotification
            receivePromotionNotification
          }
          me {
            name
          }
        }
      }
    `

    const context = {
      meLoader: () =>
        Promise.resolve({
          id: "106",
          name: "andy-warhol",
          phone: "1234890",
          location: {
            address: "123 my street",
          },
          price_range: "-1:1000000000000",
          receive_purchase_notification: true,
          receive_outbid_notification: true,
          receive_lot_opening_soon_notification: true,
          receive_sale_opening_closing_notification: true,
          receive_new_works_notification: true,
          receive_new_sales_notification: true,
          receive_promotion_notification: true,
        }),
      updateMeLoader: () =>
        Promise.resolve({
          id: "106",
          name: "andy-warhol",
          phone: "1234890",
          location: {
            address: "123 my street",
          },
          price_range: "-1:1000000000000",
          receive_purchase_notification: false,
          receive_outbid_notification: false,
          receive_lot_opening_soon_notification: false,
          receive_sale_opening_closing_notification: false,
          receive_new_works_notification: true,
          receive_new_sales_notification: false,
          receive_promotion_notification: false,
        }),
    }

    await runAuthenticatedQuery(mutation, context).then((data) => {
      expect(data).toMatchSnapshot()
    })
    expect.assertions(1)
  })
})
