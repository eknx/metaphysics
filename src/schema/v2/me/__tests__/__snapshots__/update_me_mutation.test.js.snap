// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateMeMutation updates the user profile and returns its new data payload 1`] = `
Object {
  "updateMyUserProfile": Object {
    "me": Object {
      "name": "andy-warhol",
    },
    "user": Object {
      "location": Object {
        "address": "123 my street",
        "city": null,
      },
      "name": "andy-warhol",
      "phone": "1234890",
      "priceRange": "-1:1000000000000",
    },
  },
}
`;
